### ISSUE TITLE
Customization of base_paths for relative imports

### ISSUE THREAD

**Issue by pwwang** [lark-parser/lark#603]
Currently, the `base_paths` for relative grammar imports only base on following, if I understood correctly:

1. the directory where the grammar file(-like object) is (fetching from the grammar file path)
2. the directory where the main script is (fetching from `sys.modules['__main__'].__file__`

However, this limits the architecture of projects that are using `lark`.

Imagine we have the following architecture:

```
|- package/
    |- submodule1/
        |- parser.py            # with a string grammar wanting to import some 
                                # shared rules/terminals from the shared grammar
    |- submodule2/              # maybe I have another submodule that wants to 
                                # import some rules/terminals from the shared grammar
    |- shared/
        |- grammar/
            |- shared.lark      # some shared rules/terminals
                                # say we have a terminal `INT` defined for the example below
```

Well, we could probably use 2) listed above to do the relative imports?

```lark
// not sure if this works
%import .shared.grammar.shared.INT 
```

However, this doesn't run if the package is not running independently. For example, using `pytest`, since `sys.modules['__main__'].__file__` will become the path to `pytest`.

**Ideas/Proposals:**

1. Reading paths from an env variable, say `LARK_GRAMMAR_PATH` (just like `PYTHON_PATH`)
2. Opening the `source` argument from `Lark` construct to allow overridden from outside.
3. Allowing pre-loaded grammar by `load_grammar.load_grammar`, since this function already allows specification of `grammar_name`, where the `base_paths` can fetch from

    With the above architecture, say in `parser.py`, we can do:
    ```python
    from pathlib import Path
    SHARED_GRAMMER= str(Path(__file__).parent.parent.joinpath('shared', 'grammar', 'shared.lark').resolve())
    grammar = load_grammar.load_grammar(
        r"""
        start: INT

        %import .shared.INT
        """,
        SHARED_GRAMMER
    )
    parser = Lark(grammar, ...)
    ```
    
    Then in `Lark` construct, we need to tell if the grammar is pre-loaded:
    ```python
    from load_grammer import Grammar

    class Lark:
        def __init__(self, grammar, ...):
             if isinstance(grammar, Grammar):
                 self.grammar = grammar
                 self.source = '<pre-loaded>'
                 # may have to deal with caching separately in such case
             else:
                 # regular grammar loading
    ```

**Comment by pwwang** [comment]
Just in case somebody (including me) wants a temporary solution:

Constructing a `StringIO` object and attach the path as `.name` attribute to it.

Say in `parser.py` in the above example:

```python
from io import StringIO
from pathlib import Path

COMMON_GRAMMER = str(Path(__file__).parent.parent.joinpath('common', 'grammar', 'common.lark').resolve())
grammar = StringIO(
    r"""
    start: INT

    %import .common.INT
    """
)
grammar.name = COMMON_GRAMMER 
parser = Lark(grammar, ...)
```

**Comment by erezsh** [comment]
Why not just add a `include_path` option to Lark?

Like `Lark(..., include_path='./shared/grammar')

**Comment by pwwang** [comment]
It'll be cool, too, if that can handle the nested import as well.

**Comment by bollwyvl** [comment]
Perhaps going one further, and adopting a search path, a la [jinja2](https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.FileSystemLoader) where `include_path` can either be a path or a list of paths to check in order when resolving... or actually adopt that `loader` pattern, e.g. `Lark(..., loader=SomeLoader(...))`, which given an `import` call site and the rest of the line, asks the loader for the source.

The specific use case I have in mind is a library being able to ship a "base" set of `.lark` files and allow downstream libraries/applications to extend it by customizing only the files they need, just as they might extend the visitor class the upstream exposes.

**Comment by erezsh** [comment]
I agree it should also accept a list.

I'm not sure I got what the Loader pattern is. Is it just a "readfile()" function that works on its own base path?

It's hard for me imagine the scenario you're describing. Why would a library expose its visitors?

**Comment by bollwyvl** [comment]
The specific use case is adding custom content to parsing LaTeX math
expressions in sympy... they're shopping for an antlr replacement, and I'm
partially to blame for the current implementation, so I am trying to find
an approach that eases some pain points. Pure Python and no new deps?
Check! Performance? Check (provided I can write a good grammar). At least
one other language can parse its format? Check! Tooling? Check!

One thing in particular, which this issue seems to address, is a downstream
of sympy may wish to add domain-specific meaning to the grammar (latex
being kind of... big, and not _really_ meant for this task). If there was a
simple way to:

- copy-and-update one .lark file
- add that to the loader path
- subclass whatever wrapper it would be using with one new method

Then it would be feasible to have domain-specific parsers without forking.

I haven't really used lark in anger, so
if that's a really bad idea, I'd love to know!

**Comment by erezsh** [comment]
@bollwyvl Sorry, I really don't understand your approach.

The problem, as I understand it, is that you want to define a "master grammar", and then allow small modifications on top of it, a little like subclassing.

But how will adding a loader help you do that? You can't predict which rules (or terminals) they will want to change, right?. 

It sounds more like you would want to import the grammar, and override specific rules, like you would when inheriting a class.

Am I misunderstanding?

> I haven't really used lark in anger,

Haha, I've never heard that phrase before

**Comment by pwwang** [comment]
@bollwyvl Did you mean something like this https://github.com/pwwang/liquidpy/blob/larkone/liquid/grammar.py?

I have had a very simple and basic idea implemented there to manipulate the grammars.

**Comment by bollwyvl** [comment]
I've yet to look up all of the links sent (thanks!)... but maybe won't have
to! Sometimes, the good people of the internet are just much faster than I
am:

https://github.com/sympy/sympy/pull/19825

I did raise my concerns about being able to modify behavior, but am mostly
just excited about seeing the implementation.

I think I just saw one lark file in that PR so wouldn't matter _too_ much
to this issue... but I'm still interested in how it develops!
